### DISCUSSION TITLE
How to execute a vim command after a task is finished?

### DISCUSSION THREAD

**Discussion by n0bra1n3r** [stevearc/overseer.nvim#193] [Q&A] [answered]
Is there a way to run a vim command after finishing a task? Or alternatively, is there a way to define a task that is a vim command? The specific usecase is to run a vim command to perform a hot reload on a running Flutter app after a code-generation task is finished.

**Comment by stevearc** [comment, marked as answer]
Currently tasks have to be external processes, but there are multiple ways to run a command after finishing a task.

1. You can add a subscriber to the task

```lua
task:subscribe("on_complete", function(task, status)
  if status == require("overseer").STATUS.SUCCESS then
    -- hot reload
  end
end)
```

This is very simple, but the downside is that there aren't great affordances for getting the task reference. There are [template hooks](https://github.com/stevearc/overseer.nvim/blob/master/doc/guides.md#customizing-built-in-tasks) for adjusting the _arguments_ to a task, and your custom templates also declare their own task _arguments_, but there are not as many ways to retrieve the task _reference_ once it has been launched. The other downside is that this is getting added as a function, which isn't serializable. That means that saving/loading the task to a bundle or via a session plugin will cause the callback to get lost.

2. You could create a [custom component](https://github.com/stevearc/overseer.nvim/blob/master/doc/guides.md#custom-components) called `on_complete_hot_reload` or similar. The component would define an `on_complete` handler and perform the hot reloading logic. This is a bit more boilerplate (you have to create a new file to define the component), but it works with all the serialization and is easy to add in templates, hooks, the task editor, etc.

**Reply by n0bra1n3r** [reply]
Thank you! What I ended up doing was creating a wrapper around neovim's remote API that executes vim commands inside the current vim session. It's very convenient since I can manage both vim and external tasks from one spot, and treat both task types the same way. For reference the task definition looks something like:
```lua
  return {
    args = {
      '--clean',
      '--headless',
      '--server',
      vim.v.servername,
      '--remote-expr',
      'v:lua.task_cb_runner('.._G.task_cb_id..')',
    },
    cmd = { vim.v.progpath },
    components = {
      deps,
      'on_exit_set_status',
    },
    name = name,
  }
```

**Reply by stevearc** [reply]
Oh, I wouldn't have thought of that. Really cool!
